enhancements/hypershift/node-lifecycle.md

```diff
@@ -0,0 +1,198 @@
+- Removes the token from memory after the TTL.
+
+Payload generation:
+- When generating a new payload is needed the TokenSecret controller fetches the MCO binaries for the given `.release` image.
+- It runs mco controllers in bootstrap mode.
+- It runs mcs in bootstrap mode and makes a local http query to get the payload served.
+- It stores the payload by "token" for the Ignition server (So it can be consumed by new Machines i.e. scale up or MachineDeployment rolling upgrade operations).
+- It stores the payload in a key of the token Secret (So it can be used by the Inplace upgrader i.e. upgrade of existing Machines belonging to a MachineSet).
+
+Deletion:
+- When a NodePool `.release` or `.config` is changed, a new token secret is created by the NodePool controller that corresponds to the new pair.
+- The old token Secret is marked with an expiration date by the NodePool controller.
+- Once the expiration date has passed, the old token Secret is deleted by the TokenSecret controller and its counterpart tokens in the Ignition Server memory are deleted. This strategy is done to allow in flight provisions that occurred in proximity to the nodePool upgrade to complete.
+
+#### Inplace Upgrader
+The Inplace Upgrader is a controller that runs in the control plane namespace and watches both MachineSets (management cluster) and Nodes (guest cluster).
+It orchestrates in place upgrades for a pool of Nodes i.e. MachineSets honouring the following design invariants:
+- Draining and any other step of the process but the upgrade itself are performed in a centralised fashion. This reduces the risk for high privilege Service Accounts permissions to be exposed in Nodes and therefore being vulnerable to containers escape.
+- Prevent the cluster end user as much as possible from signaling back interfering in the product owned behaviour.
+
+During the in place upgrade process as of today annotations are used to signal back and forward between NodePool<->MachineSets and between MachineSets<->Nodes.
+This might in future evolve to a different form of contract e.g. CRD.
+
+Workflow:
+- The NodePool controller signal an upgrade (`.release` and `.config` pair change) annotation the underlying MachineSet.
+- The Inplace Upgrader watches this.
+- It fetches all Nodes for the MachineSet and start the upgrade process one by one while honouring maxUnavailable.
+- Drains the Node.
+- Creates a Secret in the guest cluster with the payload content propagated from the token Secret.
+- Runs the machine-config-daemon (upgrader Pod) in the guest cluster targeted Node and mounts the secret with the payload.
+- The machine-config-daemon performs the upgrade in place and signal back state annotating the Node.
+- The Inplace Upgrader watches the Node annotation and move on appropriately, short-circuiting (in case of failure) or proceeding to next Node, until all Nodes are at the targeted version. 
```

Comment:
s/and move on/and moves on/
s/proceeding to next Node/proceeding to the next Node/